pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,600 @@
+}
+
+func (ot *ObjectTracker) setMetricsDurationInternal(i int) {
+	ot.mu.Lock()
+	defer ot.mu.Unlock()
+
+	if i > 0 && i > ot.metricsDuration {
+		ot.metricsDuration = i
+	}
+}
+
+// Call this before kubectl.Apply
+func (ot *ObjectTracker) addIfNotPresent(objects []*manifest.Object, dns string) errors.Aggregate {
+	ot.mu.Lock()
+	defer ot.mu.Unlock()
+	defer ot.deleteMetricsIfNeeded()
+	defer ot.inc()
+
+	var errs []error
+
+	for _, object := range objects {
+		gvk, ns, name := object.GroupVersionKind(),
+			object.UnstructuredObject().GetNamespace(),
+			object.UnstructuredObject().GetName()
+
+		// Check default namespace
+		if dns != "" {
+			ns = dns
+		}
+
+		namespaced, err := namespaced(ot.mgr, gvk)
+		if err != nil {
+			errs = append(errs, newNoRESTMapperErr(err, gvk))
+			continue
+		}
+
+		if namespaced {
+			if len(ns) == 0 {
+				ns = EmptyNamespace
+				errs = append(errs, newEmptyNamespaceErr(gvk, object.UnstructuredObject().GetName()))
+			}
+		} else {
+			ns = ClusterScoped
+		}
+
+		if gvkt, ok := ot.trackedGVK[gvk]; ok {
+			if !gvkt.resetIfItHas(ns, name) {
+				gvkt.insert(ns, name)
+			}
+			continue
+		}
+
+		ot.trackedGVK[gvk] = newGVKTracker(ot.mgr, object.UnstructuredObject(), namespaced)
+		ot.trackedGVK[gvk].insert(ns, name)
+
+		// addIfNotPresent is called at Reconcler.reconcileExists,
+		// so Controller & Manager is already running
+		ot.trackedGVK[gvk].start()
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+// Call after ot.mu.Lock()
+func (ot *ObjectTracker) inc() {
+	for _, gvkt := range ot.trackedGVK {
+		gvkt.list.inc()
+	}
+}
+
+// Call after ot.mu.Lock()
+func (ot *ObjectTracker) deleteMetricsIfNeeded() {
+	if ot.metricsDuration > 0 {
+		for _, gvkt := range ot.trackedGVK {
+			gvkt.deleteMetricsIfNeeded(ot.metricsDuration)
+		}
+	}
+}
+
+// NewObjectTracker returns new instance of *ObjectTracker
+func NewObjectTracker() *ObjectTracker {
+	ot := &ObjectTracker{}
+	ot.mgr = nil
+	ot.trackedGVK = make(map[schema.GroupVersionKind]*gvkTracker)
+
+	return ot
+}
+
+type noRESTMapperErr struct {
+	gvk        schema.GroupVersionKind
+	wrappedErr error
+}
+
+func (noREST noRESTMapperErr) Error() string {
+	return noREST.wrappedErr.Error()
+}
+
+func (noRESTMapperErr) Is(target error) bool {
+	_, ok := target.(noRESTMapperErr)
+	return ok
+}
+
+func newNoRESTMapperErr(err error, gvk schema.GroupVersionKind) noRESTMapperErr {
+	return noRESTMapperErr{gvk: gvk, wrappedErr: err}
+}
+
+type emptyNamespaceErr struct {
+	gvk  schema.GroupVersionKind
+	name string
+}
+
+func (emptyNamespaceErr) Error() string {
+	return "Scoped object, but no namespace specified"
+}
+
+func (emptyNamespaceErr) Is(target error) bool {
+	_, ok := target.(emptyNamespaceErr)
+	return ok
+}
+
+func newEmptyNamespaceErr(gvk schema.GroupVersionKind, name string) emptyNamespaceErr {
+	return emptyNamespaceErr{gvk: gvk, name: name}
+}
+
+type gvkTracker struct {
+	list     *items
+	src      source.Source
+	evthdler handler.EventHandler
```

Comment:
Nit: eventHandler and predicate; shorter variable names are harder to read (and probably not _that_ much faster to type any more :-) , particularly with autocomplete)